rfcs/OptInFeatures.md

```diff
@@ -0,0 +1,152 @@
+  
+
+# RFC: Opt-in features
+
+ This document is a work in progress. A lot of the questions about introspection are closely related to [#300](https://github.com/graphql/graphql-spec/issues/300) (`Expose user-defined meta-information via introspection API in form of directives`) and will therefore need to be revisited based on the progress there.
+
+## 📜 Problem Statement
+
+  
+
+GraphQL has a [built-in mechanism for deprecation](https://spec.graphql.org/draft/#sec--deprecated) allowing to gracefully remove features from the schema. The lifecycle of a feature can typically be represented as `stable` -> `deprecated` -> `removed`.  
+
+In a lot of cases though, a feature lifecycle includes an experimental phase where it has just been added and can be changed without warning. In this state, the feature is usable and feedback is encouraged but isn't considered stable enough to be put in production. The feature lifecycle becomes `experimental` -> `stable` -> `deprecated` -> `removed` .
+
+  
+### Goals
+The goal of this proposal is to support the `experimental` state and, moving forward, any state that requires the client developer to make an explicit decision before using a given feature. In that sense, it's about "opting in" to using the feature, which includes supporting `experimental` states. 
+
+To give a few examples:
+
+* a field is experimental and might be changed or removed without prior notice (the above example).
+
+* a field is expensive to compute and should be used with caution.
+
+* a field has specific security requirements and requires a specific header or other form of authentication.
+
+### Non-goals
+This proposal is not about security and/or hiding parts of a schema. Its goal is to make it easier to communicate opt-in features to client developer and therefore needs to expose that information.
+
+## 👀 Prior work
+
+* GitHub uses [schema previews](https://docs.github.com/en/graphql/overview/schema-previews) to opt-in new features.
+* Kotlin has [OptIn requirements](https://kotlinlang.org/docs/opt-in-requirements.html) that started out as `@Experimental` before [being changed to `@OptIn`](https://youtrack.jetbrains.com/issue/KT-26216/Generalize-Experimental-API)
+
+## 🧑‍💻 Proposed solution
+
+### The `@optIn` directive
+
+It is proposed to add an `@optIn` directive to the specification:
+
+```graphql
+"""
+Indicates that the given field, argument, input field or enum value requires
+giving explicit consent before being used.
+"""
+directive @optIn(feature: String!) repeatable
+    on FIELD_DEFINITION 
+    | ARGUMENT_DEFINITION 
+    | INPUT_FIELD_DEFINITION 
+    | ENUM_VALUE
```

Comment:
Could `@optIn` also apply to type definitions since you might want to expose entirely new types in an experimental feature?

Reply:
Certainly? I copy/pasted this list from `@deprecated` because I started with `@experimental` being the counterpoint of `@deprecated` but as we're moving away from this I think it makes a lot of sense. 

I guess it has implications in terms of validation though. If a type is `@optIn("foo")`, should all the usages be `optIn("foo")` too? That can make a lot of directives?

Reply:
I'm not sure about type definitions.

You can query over a union or interface, which could mean you get one of many types. Specifying what a server would have to do if you hadn't opt-ed in based _on the runtime value_ makes this much more complex. 
I think having a decidable result based only on schema & operation (and regardless of the value) is desirable.

For the same reason, I'm also not sure about `ENUM_VALUE`

ARGUMENT_DEFINITION and INPUT_FIELD_DEFINITION make sense from a static validation perspective, but I'm unsure if there's real-world need for them. 

At the field level it's often things like "this it's hitting a t1.nano and not ready for 1 million users". Experimental input fields and arguments are more likely about business logic and validation correctness. If they are truely doing something newly _experimental_ they could create a new field instead. I'd love to see some examples of situations where this would be a useful addition